src/zero_copy/mod.rs

```diff
@@ -842,6 +922,49 @@ pub trait Parser<'a, I: Input + ?Sized, O, E: ParserExtra<'a, I> = extra::Defaul
         }
     }
 
+    /// Parse one thing and then another thing, creating the second parser from the result of
+    /// the first. If you only have a couple cases to handle, prefer [`Parser::or`].
+    ///
+    /// The output of this parser is `U`, the result of the second parser
+    ///
+    /// Error recovery for this parser may be sub-optimal, as if the first parser succeeds on
+    /// recovery then the second produces an error, the primary error will point to the location in
+    /// the second parser which failed, ignoring that the first parser may be the root cause. There
+    /// may be other pathological errors cases as well.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use chumsky::zero_copy::{prelude::*, error::Simple};
+    /// // A parser that parses a single letter and then its successor
+    /// let successive_letters = one_of::<_, _, extra::Err<Simple<[u8]>>>(b'a'..=b'z')
+    ///     .then_with_ctx(
+    ///         just(b'\0').configure(|cfg, ctx: &u8| cfg.set_seq(*ctx + 1)),
+    ///         |letter, _| letter,
+    ///     );
+    ///
+    /// assert_eq!(successive_letters.parse(b"ab").into_result(), Ok(b'b')); // 'b' follows 'a'
+    /// assert!(successive_letters.parse(b"ac").has_errors()); // 'c' does not follow 'a'
+    /// ```
+    fn then_with_ctx<U, CtxN, P, F>(
+        self,
+        then: P,
+        make_ctx: F,
+    ) -> ThenWithCtx<Self, P, O, F, I, extra::Full<E::Error, E::State, CtxN>>
+    where
+        Self: Sized,
+        CtxN: 'a,
+        P: Parser<'a, I, U, extra::Full<E::Error, E::State, CtxN>>,
+        F: Fn(O, &E::Context) -> CtxN,
+    {
+        ThenWithCtx {
+            parser: self,
+            then,
+            make_ctx,
+            phantom: PhantomData,
+        }
+    }
```

Comment:
Maybe `then_with_ctx` should lose the closure and use the first parser's output as the default context? Then we could have a separate `map_ctx` method if you want more control.

New example:

```rs
one_of::<_, _, extra::Err<Simple<[u8]>>>(b'a'..=b'z')
    .then_with_ctx(just(b'\0').configure(|cfg, ctx: &u8| cfg.set_seq(*ctx + 1)))
```

Python-style indentation-sensitive parsing:

```rs
let expr = just("expr"); // TODO

let block = recursive(|block| {
    let indent = text::whitespace()
        .configure(|cfg, parent_indent| cfg.at_least(*parent_indent))
        .count();

    let expr_stmt = expr.then_ignore(text::newline());
    let control_flow = just("loop:").then(text::newline()).ignore_then(block);
    let stmt = expr_stmt.or(control_flow);

    indent.then_with_ctx(stmt)
        .repeated()
        .map(Expr::Block)
});

block.with_ctx(0) // To give an initial context for everything
```

(parses the following)

```
block:
    expr
    block:
        expr
    expr
    block:
        expr
        block:
            expr
            expr
    expr
```

Reply:
I like this - we should probably have a separate context mapping combinator anyways, and this way it nicely 'separates concerns'